diff --git a/folly/debugging/symbolizer/Elf.cpp b/folly/debugging/symbolizer/Elf.cpp
--- a/folly/debugging/symbolizer/Elf.cpp
+++ b/folly/debugging/symbolizer/Elf.cpp
@@ -43,6 +43,8 @@
 #endif
 #elif defined(__ANDROID__)
 #define FOLLY_ELF_NATIVE_CLASS __WORDSIZE
+#elif defined(__aarch64__)
+#define FOLLY_ELF_NATIVE_CLASS 64
 #endif // __ELF_NATIVE_CLASS

 namespace folly {
diff --git a/folly/memcpy_select_aarch64.cpp b/folly/memcpy_select_aarch64.cpp
--- a/folly/memcpy_select_aarch64.cpp
+++ b/folly/memcpy_select_aarch64.cpp
@@ -52,7 +52,7 @@
 #include <cstddef>
 #include <cstdint>

-#if defined(__linux__) && defined(__aarch64__)
+#if defined(__linux__) && defined(__aarch64__) && defined(__GLIBC__)

 #include <asm/hwcap.h> // @manual

diff --git a/folly/ssl/OpenSSLCertUtils.cpp b/folly/ssl/OpenSSLCertUtils.cpp
--- a/folly/ssl/OpenSSLCertUtils.cpp
+++ b/folly/ssl/OpenSSLCertUtils.cpp
@@ -31,7 +31,7 @@ std::string getOpenSSLErrorString(unsigned long err) {
   return std::string(errBuff.data());
 }

-std::string asn1ToString(ASN1_STRING* a) {
+std::string asn1ToString(const ASN1_STRING* a) {
   auto strType = ASN1_STRING_type(a);
   if (strType == V_ASN1_UTF8STRING || strType == V_ASN1_OCTET_STRING) {
     long len = ASN1_STRING_length(a);
@@ -49,9 +49,9 @@ std::string asn1ToString(ASN1_STRING* a) {
   }
 }

-std::string getExtOid(X509_EXTENSION* extension) {
+std::string getExtOid(const X509_EXTENSION* extension) {
   CHECK_NOTNULL(extension);
-  ASN1_OBJECT* object = X509_EXTENSION_get_object(extension);
+  const ASN1_OBJECT* object = X509_EXTENSION_get_object(extension);
   // Query for extension OID
   constexpr int buf_size = 256;
   std::string ret(buf_size, '\0');
@@ -65,14 +65,14 @@ std::string getExtOid(X509_EXTENSION* extension) {
   return ret;
 }

-std::string getExtData(X509_EXTENSION* extension) {
+std::string getExtData(const X509_EXTENSION* extension) {
   CHECK_NOTNULL(extension);
   auto asnValue = X509_EXTENSION_get_data(extension);
   return asnValue ? asn1ToString(asnValue) : std::string();
 }

-Optional<std::string> commonName(X509_NAME* name) {
+Optional<std::string> commonName(const X509_NAME* name) {
   if (!name) {
     return none;
   }
 
@@ -246,7 +246,7 @@ std::vector<std::string> OpenSSLCertUtils::getExtension(
     const X509& x509, folly::StringPiece oid) {
   std::vector<std::string> extValues;
   for (int i = 0; i < X509_get_ext_count(&x509); i++) {
-    X509_EXTENSION* extension = X509_get_ext(&x509, i);
+    const X509_EXTENSION* extension = X509_get_ext(&x509, i);
     std::string extensionOid = getExtOid(extension);
     if (extensionOid == oid) {
       extValues.push_back(getExtData(extension));
@@ -259,7 +259,7 @@ std::vector<std::pair<std::string, std::string>>
 OpenSSLCertUtils::getAllExtensions(const X509& x509) {
   std::vector<std::pair<std::string, std::string>> extensions;
   for (int i = 0; i < X509_get_ext_count(&x509); i++) {
-    X509_EXTENSION* extension = X509_get_ext(&x509, i);
+    const X509_EXTENSION* extension = X509_get_ext(&x509, i);
     std::string oid = getExtOid(extension);
     std::string value = getExtData(extension);
     extensions.push_back(std::make_pair(oid, value));
diff --git a/folly/io/async/ssl/OpenSSLUtils.cpp b/folly/io/async/ssl/OpenSSLUtils.cpp
--- a/folly/io/async/ssl/OpenSSLUtils.cpp
+++ b/folly/io/async/ssl/OpenSSLUtils.cpp
@@ -122,8 +122,8 @@ bool OpenSSLUtils::validatePeerCertNames(
     if ((addr4 != nullptr || addr6 != nullptr) && name->type == GEN_IPADD) {
       // Extra const-ness for paranoia
-      unsigned char const* const rawIpStr = name->d.iPAddress->data;
-      auto const rawIpLen = size_t(name->d.iPAddress->length);
+      unsigned char const* const rawIpStr = ASN1_STRING_get0_data(name->d.iPAddress);
+      auto const rawIpLen = size_t(ASN1_STRING_length(name->d.iPAddress));

       if (rawIpLen == 4 && addr4 != nullptr) {
         if (::memcmp(rawIpStr, &addr4->sin_addr, rawIpLen) == 0) {
          return true;
@@ -284,7 +284,7 @@ std::string OpenSSLUtils::getCommonName(X509* x509) {
   if (x509 == nullptr) {
     return "";
   }
-  X509_NAME* subject = X509_get_subject_name(x509);
+  const X509_NAME* subject = X509_get_subject_name(x509);
   char buf[ub_common_name + 1];
   int length =
       X509_NAME_get_text_by_NID(subject, NID_commonName, buf, sizeof(buf));
diff --git a/folly/Subprocess.cpp b/folly/Subprocess.cpp
--- a/folly/Subprocess.cpp
+++ b/folly/Subprocess.cpp
@@ -136,42 +136,42 @@ static Ret subprocess_libc_load(
 }
 
 #define FOLLY_DETAIL_SUBPROCESS_LIBC_X_BASE(X) \
-  X(_exit, _exit)                              \
-  X(close, close)                              \
-  X(dup2, dup2)                                \
-  X(fcntl, fcntl)                              \
-  X(pthread_sigmask, pthread_sigmask)          \
-  X(signal, signal)                            \
-  X(sprintf, sprintf)                          \
-  X(strtol, strtol)                            \
-  X(vfork, vfork)                              \
-  X(write, write)
+  X(_exit, _exit, &::_exit)                    \
+  X(close, close, &::close)                    \
+  X(dup2, dup2, &::dup2)                       \
+  X(fcntl, fcntl, &::fcntl)                    \
+  X(pthread_sigmask, pthread_sigmask, &::pthread_sigmask) \
+  X(signal, signal, &::signal)                 \
+  X(sprintf, sprintf, &::sprintf)              \
+  X(strtol, strtol, &::strtol)                 \
+  X(vfork, vfork, &::vfork)                    \
+  X(write, write, &::write)
 
 #if defined(__BIONIC_INCLUDE_FORTIFY_HEADERS)
 #define FOLLY_DETAIL_SUBPROCESS_LIBC_X_OPEN(X) \
-  X(open, __open_real)                         \
-  X(openat, __openat_real)
+  X(open, __open_real, static_cast<int (*)(char const*, int, ...)>(&::__open_real)) \
+  X(openat, __openat_real, static_cast<int (*)(int, char const*, int, ...)>(&::__openat_real))
 #else
 #define FOLLY_DETAIL_SUBPROCESS_LIBC_X_OPEN(X) \
-  X(open, open)                                \
-  X(openat, openat)
+  X(open, open, static_cast<int (*)(char const*, int, ...)>(&::open)) \
+  X(openat, openat, static_cast<int (*)(int, char const*, int, ...)>(&::openat))
 #endif
 
 #if defined(__linux__)
-#define FOLLY_DETAIL_SUBPROCESS_LIBC_X_PRCTL(X) X(prctl, prctl)
+#define FOLLY_DETAIL_SUBPROCESS_LIBC_X_PRCTL(X) X(prctl, prctl, &::prctl)
 #else
 #define FOLLY_DETAIL_SUBPROCESS_LIBC_X_PRCTL(X)
 #endif
@@ -181,12 +181,12 @@ static Ret subprocess_libc_load(
   FOLLY_DETAIL_SUBPROCESS_LIBC_X_OPEN(X)  \
   FOLLY_DETAIL_SUBPROCESS_LIBC_X_PRCTL(X)
 
-#define FOLLY_DETAIL_SUBPROCESS_LIBC_FIELD_DECL(name, func) \
-  static decltype(&::func) name;
-#define FOLLY_DETAIL_SUBPROCESS_LIBC_FIELD_DEFN(name, func) \
-  decltype(&::func) subprocess_libc::name;
-#define FOLLY_DETAIL_SUBPROCESS_LIBC_INIT(name, func) \
-  subprocess_libc::name = subprocess_libc_load(handle, &::func, #name);
+#define FOLLY_DETAIL_SUBPROCESS_LIBC_FIELD_DECL(name, func, ptr) \
+  static decltype(ptr) name;
+#define FOLLY_DETAIL_SUBPROCESS_LIBC_FIELD_DEFN(name, func, ptr) \
+  decltype(ptr) subprocess_libc::name;
+#define FOLLY_DETAIL_SUBPROCESS_LIBC_INIT(name, func, ptr) \
+  subprocess_libc::name = subprocess_libc_load(handle, ptr, #name);
 
 struct subprocess_libc {
   FOLLY_DETAIL_SUBPROCESS_LIBC_X(FOLLY_DETAIL_SUBPROCESS_LIBC_FIELD_DECL)
diff --git a/folly/io/async/IoUringZeroCopyBufferPool.cpp b/folly/io/async/IoUringZeroCopyBufferPool.cpp
--- a/folly/io/async/IoUringZeroCopyBufferPool.cpp
+++ b/folly/io/async/IoUringZeroCopyBufferPool.cpp
@@ -17,6 +17,7 @@
 #include <folly/io/async/IoUringZeroCopyBufferPool.h>
 
 #include <mutex>
+#include <unistd.h>
 
 #include <folly/Conv.h>
 #include <folly/String.h>
