diff --git a/bazel/gflags.bzl b/bazel/gflags.bzl
index 533fd61..1e7ad9c 100644
--- a/bazel/gflags.bzl
+++ b/bazel/gflags.bzl
@@ -49,6 +49,10 @@ def gflags_sources(namespace=["google", "gflags"]):
         "src/mutex.h",
         "src/util.h",
     ] + select({
+        "@bazel_template//bazel:cross_compiling_for_windows": [
+            "src/windows_port.cc",
+            "src/windows_port.h",
+        ],
         "//:x64_windows": [
             "src/windows_port.cc",
             "src/windows_port.h",
@@ -74,6 +78,12 @@ def gflags_library(hdrs=[], srcs=[], threads=1):
         "-DHAVE_STRTOQ",
         "-DHAVE_RWLOCK",
     ] + select({
+        "@bazel_template//bazel:cross_compiling_for_windows": [
+            "-DOS_WINDOWS",
+        ],
+        "@bazel_template//bazel:not_cross_compiling_on_windows": [
+            "-DOS_WINDOWS",
+        ],
         "//:x64_windows": [
             "-DOS_WINDOWS",
         ],
@@ -87,6 +97,7 @@ def gflags_library(hdrs=[], srcs=[], threads=1):
     if threads:
         linkopts += select({
             "//:x64_windows": [],
+            "@bazel_template//bazel:not_cross_compiling_on_windows": [],
             "//conditions:default": ["-lpthread"],
         })
     else:

