diff --git a/src/core/lib/iomgr/port.h b/src/core/lib/iomgr/port.h
--- a/src/core/lib/iomgr/port.h
+++ b/src/core/lib/iomgr/port.h
@@ -71,7 +71,11 @@
 #endif
 #define GRPC_LINUX_MULTIPOLL_WITH_EPOLL 1
 #define GRPC_POSIX_FORK 1
+#if defined(__GLIBC__)
 #define GRPC_POSIX_HOST_NAME_MAX 1
+#else
+#define GRPC_POSIX_SYSCONF 1
+#endif
 #define GRPC_POSIX_SOCKET 1
 #define GRPC_POSIX_WAKEUP_FD 1
 #ifdef __GLIBC_PREREQ
diff --git a/src/core/tsi/ssl_transport_security_utils.cc b/src/core/tsi/ssl_transport_security_utils.cc
--- a/src/core/tsi/ssl_transport_security_utils.cc
+++ b/src/core/tsi/ssl_transport_security_utils.cc
@@ -282,11 +282,11 @@ bool VerifyCrlCertIssuerNamesMatch(X509_CRL* crl, X509* cert) {
   if (cert == nullptr || crl == nullptr) {
     return false;
   }
-  X509_NAME* cert_issuer_name = X509_get_issuer_name(cert);
+  const X509_NAME* cert_issuer_name = X509_get_issuer_name(cert);
   if (cert_issuer_name == nullptr) {
     return false;
   }
-  X509_NAME* crl_issuer_name = X509_CRL_get_issuer(crl);
+  const X509_NAME* crl_issuer_name = X509_CRL_get_issuer(crl);
   if (crl_issuer_name == nullptr) {
     return false;
   }
@@ -317,7 +317,7 @@ absl::StatusOr<std::string> IssuerFromCert(X509* cert) {
   if (cert == nullptr) {
     return absl::InvalidArgumentError("cert cannot be null");
   }
-  X509_NAME* issuer = X509_get_issuer_name(cert);
+  const X509_NAME* issuer = X509_get_issuer_name(cert);
   unsigned char* buf = nullptr;
   int len = i2d_X509_NAME(issuer, &buf);
   if (len < 0 || buf == nullptr) {
@@ -334,14 +334,14 @@ absl::StatusOr<std::string> AkidFromCertificate(X509* cert) {
   if (cert == nullptr) {
     return absl::InvalidArgumentError("cert cannot be null.");
   }
-  ASN1_OCTET_STRING* akid = nullptr;
+  const ASN1_OCTET_STRING* akid = nullptr;
   int j = X509_get_ext_by_NID(cert, NID_authority_key_identifier, -1);
   // Can't have multiple occurrences
   if (j >= 0) {
     if (X509_get_ext_by_NID(cert, NID_authority_key_identifier, j) != -1) {
       return absl::InvalidArgumentError("Could not get AKID from certificate.");
     }
-    akid = X509_EXTENSION_get_data(X509_get_ext(cert, j));
+    akid = X509_EXTENSION_get_data(X509_get_ext(cert, j));
   } else {
     return absl::InvalidArgumentError("Could not get AKID from certificate.");
   }
@@ -359,14 +359,14 @@ absl::StatusOr<std::string> AkidFromCrl(X509_CRL* crl) {
   if (crl == nullptr) {
     return absl::InvalidArgumentError("Could not get AKID from crl.");
   }
-  ASN1_OCTET_STRING* akid = nullptr;
+  const ASN1_OCTET_STRING* akid = nullptr;
   int j = X509_CRL_get_ext_by_NID(crl, NID_authority_key_identifier, -1);
   // Can't have multiple occurrences
   if (j >= 0) {
     if (X509_CRL_get_ext_by_NID(crl, NID_authority_key_identifier, j) != -1) {
       return absl::InvalidArgumentError("Could not get AKID from crl.");
     }
-    akid = X509_EXTENSION_get_data(X509_CRL_get_ext(crl, j));
+    akid = X509_EXTENSION_get_data(X509_CRL_get_ext(crl, j));
   } else {
     return absl::InvalidArgumentError("Could not get AKID from crl.");
   }
diff --git a/src/core/tsi/ssl_transport_security.cc b/src/core/tsi/ssl_transport_security.cc
--- a/src/core/tsi/ssl_transport_security.cc
+++ b/src/core/tsi/ssl_transport_security.cc
@@ -891,9 +891,9 @@ static tsi_result ssl_get_x509_common_name(X509* cert, unsigned char** utf8,
                                            size_t* utf8_size) {
   int common_name_index = -1;
-  X509_NAME_ENTRY* common_name_entry = nullptr;
-  ASN1_STRING* common_name_asn1 = nullptr;
-  X509_NAME* subject_name = X509_get_subject_name(cert);
+  const X509_NAME_ENTRY* common_name_entry = nullptr;
+  const ASN1_STRING* common_name_asn1 = nullptr;
+  const X509_NAME* subject_name = X509_get_subject_name(cert);
   int utf8_returned_size = 0;
   if (subject_name == nullptr) {
     VLOG(2) << "Could not get subject name from certificate.";
@@ -951,7 +951,7 @@ static tsi_result peer_property_from_x509_subject(X509* cert,
                                                   tsi_peer_property* property,
                                                   bool is_verified_root_cert) {
-  X509_NAME* subject_name = X509_get_subject_name(cert);
+  const X509_NAME* subject_name = X509_get_subject_name(cert);
   if (subject_name == nullptr) {
     GRPC_TRACE_LOG(tsi, INFO) << "Could not get subject name from certificate.";
     return TSI_NOT_FOUND;
@@ -1047,18 +1047,20 @@ static tsi_result add_subject_alt_names_properties_to_peer(
     } else if (subject_alt_name->type == GEN_IPADD) {
       char ntop_buf[INET6_ADDRSTRLEN];
       int af;
+      const ASN1_OCTET_STRING* ip_address = subject_alt_name->d.iPAddress;
+      const int ip_address_length = ASN1_STRING_length(ip_address);
 
-      if (subject_alt_name->d.iPAddress->length == 4) {
+      if (ip_address_length == 4) {
         af = AF_INET;
-      } else if (subject_alt_name->d.iPAddress->length == 16) {
+      } else if (ip_address_length == 16) {
         af = AF_INET6;
       } else {
         LOG(ERROR) << "SAN IP Address contained invalid IP";
         result = TSI_INTERNAL_ERROR;
         break;
       }
-      const char* name = inet_ntop(af, subject_alt_name->d.iPAddress->data,
-                                   ntop_buf, INET6_ADDRSTRLEN);
+      const char* name = inet_ntop(af, ASN1_STRING_get0_data(ip_address),
+                                   ntop_buf, INET6_ADDRSTRLEN);
       if (name == nullptr) {
         LOG(ERROR) << "Could not get IP string from asn1 octet.";
         result = TSI_INTERNAL_ERROR;
@@ -1338,17 +1340,17 @@ static tsi_result x509_store_load_certs(X509_STORE* cert_store,
   while (true) {
     root = PEM_read_bio_X509_AUX(pem, nullptr, nullptr, const_cast<char*>(""));
     if (root == nullptr) {
       ERR_clear_error();
       break;  // We're at the end of stream.
     }
     if (root_names != nullptr) {
-      root_name = X509_get_subject_name(root);
-      if (root_name == nullptr) {
+      const X509_NAME* subject_name = X509_get_subject_name(root);
+      if (subject_name == nullptr) {
         LOG(ERROR) << "Could not get name from root certificate.";
         result = TSI_INVALID_ARGUMENT;
         break;
       }
-      root_name = X509_NAME_dup(root_name);
+      root_name = X509_NAME_dup(subject_name);
       if (root_name == nullptr) {
         result = TSI_OUT_OF_RESOURCES;
         break;
diff --git a/src/core/credentials/transport/tls/grpc_tls_crl_provider.cc b/src/core/credentials/transport/tls/grpc_tls_crl_provider.cc
--- a/src/core/credentials/transport/tls/grpc_tls_crl_provider.cc
+++ b/src/core/credentials/transport/tls/grpc_tls_crl_provider.cc
@@ -56,7 +56,7 @@ absl::StatusOr<std::string> IssuerFromCrl(X509_CRL* crl) {
   if (crl == nullptr) {
     return absl::InvalidArgumentError("crl cannot be null");
   }
-  X509_NAME* issuer = X509_CRL_get_issuer(crl);
+  const X509_NAME* issuer = X509_CRL_get_issuer(crl);
   if (issuer == nullptr) {
     return absl::InvalidArgumentError("crl cannot have null issuer");
   }
