diff --git a/proxygen/lib/http/HTTPPriorityFunctions.cpp b/proxygen/lib/http/HTTPPriorityFunctions.cpp
--- a/proxygen/lib/http/HTTPPriorityFunctions.cpp
+++ b/proxygen/lib/http/HTTPPriorityFunctions.cpp
@@ -48,8 +48,8 @@
   bool logBadHeader = false;
   SCOPE_EXIT {
     if (logBadHeader) {
-      LOG_EVERY_N(ERROR, 100)
-          << "Received ill-formated priority header=" << priority;
+      // LOG_EVERY_N(ERROR, 100)
+      //     << "Received ill-formated priority header=" << priority;
     }
   };
   StructuredHeadersDecoder decoder(priority);
diff --git a/proxygen/lib/http/structuredheaders/StructuredHeadersBuffer.cpp b/proxygen/lib/http/structuredheaders/StructuredHeadersBuffer.cpp
--- a/proxygen/lib/http/structuredheaders/StructuredHeadersBuffer.cpp
+++ b/proxygen/lib/http/structuredheaders/StructuredHeadersBuffer.cpp
@@ -284,10 +284,10 @@
 }
 
 DecodeError StructuredHeadersBuffer::handleDecodeError(const DecodeError& err) {
-  LOG_EVERY_N(ERROR, 1000)
-      << "Error message: " << decodeErrToString(err)
-      << ". Number of characters parsed before error:" << getNumCharsParsed()
-      << ". Header Content:" << originalContent_.str();
+  // LOG_EVERY_N(ERROR, 1000)
+  //     << "Error message: " << decodeErrToString(err)
+  //     << ". Number of characters parsed before error:" << getNumCharsParsed()
+  //     << ". Header Content:" << originalContent_.str();
   return err;
 }
 
diff --git a/proxygen/lib/http/structuredheaders/StructuredHeadersEncoder.cpp b/proxygen/lib/http/structuredheaders/StructuredHeadersEncoder.cpp
--- a/proxygen/lib/http/structuredheaders/StructuredHeadersEncoder.cpp
+++ b/proxygen/lib/http/structuredheaders/StructuredHeadersEncoder.cpp
@@ -223,14 +223,14 @@
 // was involved in the error
 EncodeError StructuredHeadersEncoder::handleEncodeError(
     EncodeError err, const std::string& culprit) {
-  LOG_EVERY_N(ERROR, 1000) << "Error message: " << encodeErrToString(err)
-                           << ", culprit: " << culprit;
+  // LOG_EVERY_N(ERROR, 1000) << "Error message: " << encodeErrToString(err)
+  //                          << ", culprit: " << culprit;
   return err;
 }
 
 // Used to print more general error messages (eg: empty data structure)
 EncodeError StructuredHeadersEncoder::handleEncodeError(EncodeError err) {
-  LOG_EVERY_N(ERROR, 1000) << "Error message: " << encodeErrToString(err);
+  // LOG_EVERY_N(ERROR, 1000) << "Error message: " << encodeErrToString(err);
   return err;
 }
 
diff --git a/proxygen/lib/services/RequestWorkerThread.cpp b/proxygen/lib/services/RequestWorkerThread.cpp
--- a/proxygen/lib/services/RequestWorkerThread.cpp
+++ b/proxygen/lib/services/RequestWorkerThread.cpp
@@ -53,6 +53,7 @@
 void RequestWorkerThread::setup() {
   CHECK(evb_);
   evb_->runImmediatelyOrRunInEventBaseThreadAndWait([&]() {
+#if !defined(_MSC_VER) && !defined(__MINGW32__)
     sigset_t ss;
 
     // Ignore some signals
@@ -68,7 +69,7 @@
     sigaddset(&ss, SIGCHLD);
     sigaddset(&ss, SIGIO);
     PCHECK(pthread_sigmask(SIG_BLOCK, &ss, nullptr) == 0);
-
+#endif
     currentRequestWorker_ = this;
     callback_.workerStarted(this);
   });
diff --git a/proxygen/lib/services/WorkerThread.cpp b/proxygen/lib/services/WorkerThread.cpp
--- a/proxygen/lib/services/WorkerThread.cpp
+++ b/proxygen/lib/services/WorkerThread.cpp
@@ -13,6 +13,7 @@
 #include <folly/String.h>
 #include <folly/io/async/EventBaseManager.h>
 #include <folly/io/async/IoUringBackend.h>
+#include <gflags/gflags.h>
 #include <glog/logging.h>
 #include <signal.h>
 
@@ -150,7 +151,7 @@
 }
 
 void WorkerThread::setup() {
-#ifndef _MSC_VER
+#if !defined(_MSC_VER) && !defined(__MINGW32__)
   sigset_t ss;
 
   // Ignore some signals
diff --git a/proxygen/lib/transport/H3DatagramAsyncSocket.cpp b/proxygen/lib/transport/H3DatagramAsyncSocket.cpp
--- a/proxygen/lib/transport/H3DatagramAsyncSocket.cpp
+++ b/proxygen/lib/transport/H3DatagramAsyncSocket.cpp
@@ -167,7 +167,7 @@
       // buffer until reads are resumed
       readBuf_.emplace_back(std::move(datagram));
     } else {
-      VLOG_EVERY_N(2, 1000) << "Dropped incoming datagram.";
+      // VLOG_EVERY_N(2, 1000) << "Dropped incoming datagram.";
     }
     return;
   }
diff --git a/proxygen/lib/utils/StateMachine.h b/proxygen/lib/utils/StateMachine.h
--- a/proxygen/lib/utils/StateMachine.h
+++ b/proxygen/lib/utils/StateMachine.h
@@ -30,9 +30,9 @@
 
     std::tie(newState, ok) = T::find(state, event);
     if (!ok) {
-      LOG_EVERY_N(ERROR, 100)
-          << T::getName() << ": invalid transition tried: " << state << " "
-          << event;
+      // LOG_EVERY_N(ERROR, 100)
+      //     << T::getName() << ": invalid transition tried: " << state << " "
+      //     << event;
       return false;
     } else {
       VLOG(6) << T::getName() << ": transitioning from " << state << " to "
