diff --git a/boost/boost.bzl b/boost/boost.bzl
index 2f6b0c3..4fb1176 100644
--- a/boost/boost.bzl
+++ b/boost/boost.bzl
@@ -1,3 +1,4 @@
+load("@rules_cc//cc:defs.bzl", "cc_binary", "cc_import", "cc_library")
 load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 load("@bazel_tools//tools/build_defs/repo:utils.bzl", "maybe")
 
@@ -37,7 +38,7 @@ def boost_library(
     if boost_name == None:
         boost_name = name
 
-    return native.cc_library(
+    return cc_library(
         name = name,
         visibility = visibility,
         defines = default_defines + defines,
@@ -70,7 +71,7 @@ def boost_so_library(
     if boost_name == None:
         boost_name = name
 
-    native.cc_binary(
+    cc_binary(
         name = "lib_internal_%s" % name,
         defines = default_defines + defines,
         includes = ["libs/%s/include" % boost_name] + includes,
@@ -87,7 +88,7 @@ def boost_so_library(
         output_group = "interface_library",
         visibility = ["//visibility:private"],
     )
-    native.cc_import(
+    cc_import(
         name = "_imported_%s" % name,
         shared_library = ":lib_internal_%s" % name,
         interface_library = ":%s_dll_interface_file" % name,
